test/fetch-server.js

```diff
@@ -1,5 +1,156 @@
+import {FetchServer} from '../lib';
+import chai, {expect} from 'chai';
+
+const httpMethodsHandler = {
+  '/some/route': {
```

Comment:
My 2 cents of feedback. I fully acknowledge you've spent a ton of effort into this already and I appreciate it.

1) Right now the interface is that `FetchServer` class is exported, but the server itself is not stateful, nor is it possible to have multiple concurrent instances of FetchServer running. Instead how about we only expose two methods. so it's used as following

```ts
import {mockFetch, restoreFetch) from 'mock-fetch';

beforeEach(function() {
 mockFetch(routesConfig);
});

afterEach(function() {
  restoreFetch();
});

```

2) typescript please, seeing the concrete definition of the interface is easier to work with. I know this was already in js but types help consume the interface.

I see the interface allows returning objects like 

```
GET: {
      'content-type': 'application/json',
       'status': 200,
        'result': {}
      },
```

but that feels weird, because window.fetch signature is `fetch(url: Request | string, init?: RequestInit): Promise<Response>`. We should embrace the Response interface. The response constructor allows passing of headers and custom response types, but it's flexible enough to handle all sorts of other responses like csv/binary blobs/form data, status codes e.t.c https://developer.mozilla.org/en-US/docs/Web/API/Response/Response.

I understand 99% use case is we're returning json objects or arrays with status 200 and a shorthand for that should be nice, but I think that's all we should handle, and let the user fallback to using `new Response()` for any crazy custom response they may want to respond with.

Thoughts on the following proposed interface based on what you already have here?

```ts

interface MockFetchOptions {
  /* console logs requests and responses */
  debug?: boolean;
}

/* Pojo = Plain old javascript object - safe for JSON.stringify() */
type Pojo = string | number | boolean | Record<string, any> | any[];
type FetchResponder = (url: string, options: RequestInit) => Response | Pojo | Promise<Response | Pojo>;
type FetchMethodResponder = {
  GET?: FetchResponder,
  POST?: FetchResponder,
  PUT?: FetchResponder,
  PATCH?: FetchResponder,
  DELETE?: FetchResponder,
}
type FetchRoute = FetchMethodResponder | FetchResponder;
type FetchRoutes = Record<string, FetchRoute>;

/** handles urls matched by routes, any unhandled urls will throw an error.  '*' is a special route that handles any urls that don't mach */
export function mockFetch(routes: FetchRoutes, options?: MockFetchOptions): void;
export function restoreFetch(): void;
```


3) You're right, domsuite/index.js doesn't depend on sinon, only fetch-server does. If we mock fetch server like other places we override fetch, then we don't need sinon in all of domsuite. The package becomes 99% smaller.

```ts
let _fetchRoutes: FetchRoutes;

export function mockFetch(routes: FetchRoutes, options?: MockFetchOptions): void {
  _fetchRoutes = routes;
  if (window.__origFetch]) {
    // already mocked, 
    return;
  }

  window.__origFetch = window.fetch;
  window.fetch = function mockFetch(url: string, options: RequestInit) {
    ...
  };
}

export function restoreFetch(): void {
 _fetchRoutes = null;
 if (window.__origFetch]) {
    window.fetch = window.__origFetch;
    delete window.__origFetch;
 }
}
```


This would let us fully mock out api-client's fetch-server and resolve design-system's use-case too.

In fact if we have a shared library of app api mocks, we can have proper visual regression testing that is network independent.

Reply:
haha, this `mockFetch` and `restoreFetch` is literally the interface isorender exposes:
https://github.com/mixpanel/analytics/blob/master/iron/common/test/isorender/utils.ts#L973-L1025
Seems like we may be onto something...